File: NoughtsAndCrosses/Utils/DataBuffer.cs
using System;
using System.IO;
using System.Text;

namespace NoughtsAndCrosses.Utils {
  public sealed class DataBuffer {

    #region Инициализация

    private byte[] buffer;
    private uint bufferSize;
    private uint size;

    public DataBuffer(int aCapacity) {
      buffer = new byte[aCapacity];
      bufferSize = 0;
    }

    public DataBuffer() : this(1) {
    }

    public DataBuffer(byte[] data, uint size) {
      this.buffer = data;
      this.size = size;
    }

    public DataBuffer(string str) : this(str.Length) {
      SetAt(0, Encoding.Default.GetBytes(str));
    }

    public uint GetBufferSize() {
      return bufferSize;
    }

    public void Clear() {
      bufferSize = 0;
    }

    public byte[] GetBufferCopy() {
      byte[] bufferCopy = new byte[bufferSize];
      lock (buffer)
        Array.Copy(buffer, 0, bufferCopy, 0, bufferSize);
      return bufferCopy;
    }

    public byte[] GetBuffer() {
      return buffer;
    }

    private void Grow(uint aCapacity) {
      if (buffer.Length < aCapacity) {
        uint newSize = (uint)Math.Max(aCapacity, buffer.Length * 2);
        byte[] newBuffer = new byte[newSize];
        Array.Copy(buffer, 0, newBuffer, 0, buffer.Length);
        buffer = newBuffer;
      }
    }
    #endregion

    #region Добавление в конец

    public void Add(byte[] data, uint dataOffset, uint aSize) {
      SetAt(bufferSize, data, dataOffset, aSize);
    }

    public void Add(byte[] data, uint aSize) {
      SetAt(bufferSize, data, 0, aSize);
    }

    public void Add(byte[] data) {
      SetAt(bufferSize, data, 0, (uint)data.Length);
    }

    public void Add(byte val) {
      SetAt(bufferSize, val);
    }

    public void Add(bool val) {
      SetAt(bufferSize, val);
    }

    public void Add(char val) {
      SetAt(bufferSize, BitConverter.GetBytes(val));
    }

    public void Add(short val) {
      SetAt(bufferSize, BitConverter.GetBytes(val));
    }

    public void Add(ushort val) {
      SetAt(bufferSize, BitConverter.GetBytes(val));
    }

    public void Add(int val) {
      SetAt(bufferSize, BitConverter.GetBytes(val));
    }

    public void Add(uint val) {
      SetAt(bufferSize, BitConverter.GetBytes(val));
    }

    public void Add(long val) {
      SetAt(bufferSize, BitConverter.GetBytes(val));
    }

    public void Add(ulong val) {
      SetAt(bufferSize, BitConverter.GetBytes(val));
    }

    public void Add(float val) {
      SetAt(bufferSize, BitConverter.GetBytes(val));
    }

    public void Add(double val) {
      SetAt(bufferSize, BitConverter.GetBytes(val));
    }

    public void AddUnicode(string str) {
      SetAtUnicode(bufferSize, str);
    }

    public void AddLargeUnicode(string str) {
      SetAtLargeUnicode(bufferSize, str);
    }

    public void AddASCII(string str) {
      SetAtASCII(bufferSize, str);
    }

    public void AddLargeASCII(string str) {
      SetAtLargeASCII(bufferSize, str);
    }

    public void AddArray(byte[] data, uint dataOffset, ushort aSize) {
      SetAtArray(bufferSize, data, dataOffset, aSize);
    }

    public void AddArray(byte[] data) {
      SetAtArray(bufferSize, data, 0, (ushort)data.Length);
    }

    public void AddLargeArray(byte[] data) {
      SetAtArray(bufferSize, data, 0, (uint)data.Length);
    }

    public void AddLargeArray(byte[] data, uint dataOffset, uint aSize) {
      SetAtArray(bufferSize, data, dataOffset, aSize);
    }

    public void AddData(DataBuffer data) {
      SetAtArray(bufferSize, data.GetBuffer(), 0, data.GetBufferSize());
    }

    public bool AddData(BinaryReader ins) {
      return SetAt(bufferSize, ins);
    }

    public void Truncate(uint offset) {
      bufferSize = Math.Min(bufferSize, offset);
    }
    #endregion

    #region Вставка по заданному смещению

    public void SetAt(uint offset, byte[] data, uint dataOffset, uint aSize) {
      lock (buffer) {
        Grow(offset + aSize);
        Array.Copy(data, dataOffset, buffer, offset, aSize);
        bufferSize = Math.Max(bufferSize, offset + aSize);
      }
    }

    public void SetAt(uint offset, byte[] data, uint aSize) {
      SetAt(offset, data, 0, aSize);
    }

    public void SetAt(uint offset, byte[] data) {
      SetAt(offset, data, 0, (uint)data.Length);
    }

    public void SetAt(uint offset, byte val) {
      lock (buffer) {
        Grow(offset + 1);
        buffer[offset] = val;
        bufferSize = Math.Max(bufferSize, offset + 1);
      }
    }

    public void SetAt(uint offset, bool val) {
      SetAt(offset, BitConverter.GetBytes(val)) ;
    }

    public void SetAt(uint offset, char val) {
      SetAt(offset, BitConverter.GetBytes(val));
    }

    public void SetAt(uint offset, short val) {
      SetAt(offset, BitConverter.GetBytes(val));
    }

    public void SetAt(uint offset, ushort val) {
      SetAt(offset, BitConverter.GetBytes(val));
    }

    public void SetAt(uint offset, int val) {
      SetAt(offset, BitConverter.GetBytes(val));
    }

    public void SetAt(uint offset, uint val) {
      SetAt(offset, BitConverter.GetBytes(val));
    }

    public void SetAt(uint offset, long val) {
      SetAt(offset, BitConverter.GetBytes(val));
    }

    public void SetAt(uint offset, ulong val) {
      SetAt(offset, BitConverter.GetBytes(val));
    }

    public void SetAt(uint offset, float val) {
      SetAt(offset, BitConverter.GetBytes(val));
    }

    public void SetAt(uint offset, double val) {
      SetAt(offset, BitConverter.GetBytes(val));
    }

    public void SetAtUnicode(uint offset, string str) {
      lock (buffer) {
        ushort len = (ushort)str.Length;
        SetAt(offset, BitConverter.GetBytes(len));
        if (str.Length == 0) {
          return;
        }
        SetAt(offset + 2, Encoding.Unicode.GetBytes(str));
      }
    }

    public void SetAtLargeUnicode(uint offset, string str) {
      lock (buffer) {
        SetAt(offset, BitConverter.GetBytes(str.Length));
        if (str.Length == 0) {
          return;
        }
        SetAt(offset + 4, Encoding.Unicode.GetBytes(str));
      }
    }

    public void SetAtASCII(uint offset, string str) {
      lock (buffer) {
        ushort len = (ushort)str.Length;
        SetAt(offset, BitConverter.GetBytes(len));
        if (str.Length == 0) {
          return;
        }
        SetAt(offset + 2, Encoding.Default.GetBytes(str));
      }
    }

    public void SetAtLargeASCII(uint offset, string str) {
      lock (buffer) {
        SetAt(offset, BitConverter.GetBytes(str.Length));
        if (str.Length == 0) {
          return;
        }
        SetAt(offset + 4, Encoding.Default.GetBytes(str));
      }
    }

    public void SetAtArray(uint offset, byte[] data, uint dataOffset, ushort aSize) {
      lock (buffer) {
        SetAt(offset, aSize);
        if (aSize == 0) {
          return;
        }
        SetAt(offset + 2, data, dataOffset, aSize);
      }
    }

    public void SetAtArray(uint offset, byte[] data) {
      SetAtArray(offset, data, 0, (ushort)data.Length);
    }

    public void SetAtArray(uint offset, byte[] data, uint dataOffset, uint aSize) {
      lock (buffer) {
        SetAt(offset, aSize);
        if (aSize == 0) {
          return;
        }
        SetAt(offset + 4, data, dataOffset, aSize);
      }
    }

    public bool SetAt(uint offset, BinaryReader ins) {
      try {
        uint i = 0;
        int c = 0;
        while (true) {
          Grow(offset + i + 1);
          c = ins.Read();
          if (c == -1) {
            break;
          }
          buffer[offset + i] = Convert.ToByte(c);// ins.ReadByte();
          ++i;
        }
        bufferSize = Math.Max(bufferSize, offset + i);
        return true;
      }
      catch (IOException ex) {
        return false;
      }
    }

    #endregion

  }
}

File: NCServer/src/ncserver/utils/DataBuffer.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package ncserver.utils;

import java.io.IOException;
import java.io.InputStream;

/**
 *
 * @author Vladimir
 */
public final class DataBuffer {

    // <editor-fold defaultstate="collapsed" desc="Инициализация">
    protected byte[] buffer;
    protected int bufferSize;

    public DataBuffer(int capacity) {
        buffer = new byte[capacity];
        bufferSize = 0;
    }

    public DataBuffer() {
        buffer = new byte[1];
        bufferSize = 0;
    }

    public DataBuffer(String str) {
        this(str.length());
        setAt(0, BitConverter.getBytes(str));
    }

    public int getBufferSize() {
        return bufferSize;
    }

    public void clear() {
        bufferSize = 0;
    }

    public byte[] getBufferCopy() {
        byte[] bufferCopy = new byte[bufferSize];
        synchronized (buffer) {
            System.arraycopy(buffer, 0, bufferCopy, 0, bufferSize);
        }
        return bufferCopy;
    }

    public byte[] getBuffer() {
        return buffer;
    }

    protected void grow(int capacity) {
        if (buffer.length < capacity) {
            int newSize = Math.max(capacity, buffer.length * 2);
            byte[] newBuffer = new byte[newSize];
            System.arraycopy(buffer, 0, newBuffer, 0, buffer.length);
            buffer = newBuffer;
        }
    }
    // </editor-fold>

    // <editor-fold defaultstate="collapsed" desc="Добавление в конец">
    public void add(byte[] data, int dataOffset, int size) {
        setAt(bufferSize, data, dataOffset, size);
    }

    public void add(byte[] data, int size) {
        setAt(bufferSize, data, 0, size);
    }

    public void add(byte[] data) {
        setAt(bufferSize, data, 0, (int)data.length);
    }

    public void add(byte number) {
        setAt(bufferSize, number);
    }

    public void add(boolean number) {
        setAt(bufferSize, number);
    }

    public void add(char number) {
        setAt(bufferSize, BitConverter.getBytes(number));
    }

    public void add(short number) {
        setAt(bufferSize, BitConverter.getBytes(number));
    }

    public void add(int number) {
        setAt(bufferSize, BitConverter.getBytes(number));
    }

    public void add(long number) {
        setAt(bufferSize, BitConverter.getBytes(number));
    }

    public void add(float number) {
        setAt(bufferSize, BitConverter.getBytes(number));
    }

    public void add(double number) {
        setAt(bufferSize, BitConverter.getBytes(number));
    }

    public void addUnicode(String str) {
        setAtUnicode(bufferSize, str);
    }

    public void addLargeUnicode(String str) {
        setAtLargeUnicode(bufferSize, str);
    }

    public void addASCII(String str) {
        setAtASCII(bufferSize, str);
    }

    public void addLargeASCII(String str) {
        setAtLargeASCII(bufferSize, str);
    }

    public void addArray(byte[] data, int dataOffset, short size) {
        setAtArray(bufferSize, data, dataOffset, size);
    }

    public void addArray(byte[] data) {
        setAtArray(bufferSize, data, 0, (short)data.length);
    }

    public void addLargeArray(byte[] data) {
        setAtArray(bufferSize, data, 0, (int)data.length);
    }

    public void addLargeArray(byte[] data, int dataOffset, int size) {
        setAtArray(bufferSize, data, dataOffset, size);
    }

    public void addData(DataBuffer data) {
        setAtArray(bufferSize, data.getBuffer(), 0, data.getBufferSize());
    }

    public boolean addData(InputStream ins) {
        return setAt(bufferSize, ins);
    }

    public void truncate(int offset) {
        bufferSize = Math.min(bufferSize, offset);
    }
    // </editor-fold>

    // <editor-fold defaultstate="collapsed" desc="Вставка по заданному смещению">
    public void setAt(int offset, byte[] data, int dataOffset, int size) {
        synchronized (buffer) {
            grow(offset + size);
            System.arraycopy(data, dataOffset, buffer, offset, size);
            bufferSize = Math.max(bufferSize, offset + size);
        }
    }

    public void setAt(int offset, byte[] data, int size) {
        setAt(offset, data, 0, size);
    }

    public void setAt(int offset, byte[] data) {
        setAt(offset, data, 0, (int)data.length);
    }

    public void setAt(int offset, byte number) {
        synchronized (buffer) {
            grow(offset + 1);
            buffer[offset] = number;
            bufferSize = Math.max(bufferSize, offset + 1);
        }
    }

    public void setAt(int offset, boolean number) {
        setAt(offset, BitConverter.getBytes(number));
    }

    public void setAt(int offset, char number) {
        setAt(offset, BitConverter.getBytes(number));
    }

    public void setAt(int offset, short number) {
        setAt(offset, BitConverter.getBytes(number));
    }

    public void setAt(int offset, int number) {
        setAt(offset, BitConverter.getBytes(number));
    }

    public void setAt(int offset, long number) {
        setAt(offset, BitConverter.getBytes(number));
    }

    public void setAt(int offset, float number) {
        setAt(offset, BitConverter.getBytes(number));
    }

    public void setAt(int offset, double number) {
        setAt(offset, BitConverter.getBytes(number));
    }

    public void setAtUnicode(int offset, String str) {
        synchronized (buffer) {
            short len = (short)str.length();
            setAt(offset, BitConverter.getBytes(len));
            if (str.length() == 0) {
                return;
            }
            setAt(offset + 2, BitConverter.getBytes(str));
        }
    }

    public void setAtLargeUnicode(int offset, String str) {
        synchronized (buffer) {
            setAt(offset, BitConverter.getBytes(str.length()));
            if (str.length() == 0) {
                return;
            }
            setAt(offset + 4, BitConverter.getBytes(str));
        }
    }

    public void setAtASCII(int offset, String str) {
        synchronized (buffer) {
            short len = (short)str.length();
            setAt(offset, BitConverter.getBytes(len));
            if (str.length() == 0) {
                return;
            }
            setAt(offset + 2, BitConverter.getBytes(str));
        }
    }

    public void setAtLargeASCII(int offset, String str) {
        synchronized (buffer) {
            setAt(offset, BitConverter.getBytes(str.length()));
            if (str.length() == 0) {
                return;
            }
            setAt(offset + 4, BitConverter.getBytes(str));
        }
    }

    public void setAtArray(int offset, byte[] data, int dataOffset, short size) {
        synchronized (buffer) {
            setAt(offset, size);
            if (size == 0) {
                return;
            }
            setAt(offset + 2, data, dataOffset, size);
        }
    }

    public void setAtArray(int offset, byte[] data) {
        setAtArray(offset, data, 0, (short)data.length);
    }

    public void setAtArray(int offset, byte[] data, int dataOffset, int size) {
        synchronized (buffer) {
            setAt(offset, size);
            if (size == 0) {
                return;
            }
            setAt(offset + 4, data, dataOffset, size);
        }
    }

    public boolean setAt(int offset, InputStream ins) {
        try {
            int i = 0, c = 0;
            while (true) {
                grow(offset + i + 1);
                c = ins.read();
                if (c == -1) {
                    break;
                }
                buffer[offset + i] = (byte)c;
                ++i;
            }
            bufferSize = Math.max(bufferSize, offset + i);
            return true;
        }
        catch (IOException ex) {
            System.err.println(ex);
            return false;
        }
    }

    // </editor-fold>
}
